### ISSUE TITLE
Grails 3.0.9: Unable to create an app without internet access (ie in offline mode)

### ISSUE THREAD

**Issue by weberdc** [grails/grails-core#9393]
Related to Issue #594, I want to create an app on a network not connected to the Internet or a standalone machine using Grails 3.0.9. When I tell it to use offline mode, it still seems to want to check the Internet for project profiles - could a default profile could be baked in?

The output starts as follows:
<pre>
[derek@myhost]$ grails --offline --stacktrace create-app test_app
| Error Problem updating profiles from origin git repository (NOTE: Stack trace has been filtered. Use --verbose to see entire trace.)
org.eclipse.jgit.api.errors.TransportException: https://github.com/grail/grails-profile-repository: cannot open git-upload-pack
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:139)
        ...
</pre>

I expected that the `--offline` flag would be enough to get grails to just muddle through with what resources it comes with out of the box, but I take it that's not the case.
I'm running grails 3.0.9, groovy 2.4.5 with JDK1.8.0_45 on Centos 6.5.

**Comment by weberdc** [comment]
In addition to this, I created a grails project on an Internet-connected PC, let it look for profiles on github and download a huge number of libraries into my ~/.gradle directory to run the grails shell, and then zipped up both the project directory and the ~/.gradle directory, and brought it all to the disconnected network. Now when I try to run `grails --offline` within the project, it still attempts to find the grails-gradle-plugin from the Internet, so it's clear that the `--offline` flag doesn't just not work as expected when creating apps, but also when running standard grails commands too.

**Comment by Luxor** [comment]
Yeah, I have the same issue. I have connection to internal maven repository only (which has internet connection). 

Can I override some settings to force grails use my repository on create-app command? E.g. gradle default repository.

Please advice

**Comment by Luxor** [comment]
Looks like grails i trying to connect to 
```
Caused by: org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact org.grails.profiles:angular:pom:3.1.4 from/to grailsCentral (https://repo.grails.org/grails/core): Conn
```
Can I override this grails repository and use internal one, eg. my.corporate.com?

At the end it failed to load a profile and stops.
```
Error |
Failed to load latest version of profile [angular]. Trying Grails release version
org.eclipse.aether.resolution.DependencyResolutionException: Failed to collect dependencies at org.grails.profiles:angular:jar:3.1.4
Error |
Error occurred running Grails CLI: connect timed out
java.net.SocketTimeoutException: connect timed out
        at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method)
        at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:85)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:345)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:656)
        at org.apache.http.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:414)
        at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
        at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:294)
        at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:643)
        at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:479)
        at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:906)
        at org.apache.http.impl.client.DecompressingHttpClient.execute(DecompressingHttpClient.java:137)
        at org.eclipse.aether.transport.http.HttpTransporter.execute(HttpTransporter.java:287)
        at org.eclipse.aether.transport.http.HttpTransporter.implGet(HttpTransporter.java:243)
        at org.eclipse.aether.spi.connector.transport.AbstractTransporter.get(AbstractTransporter.java:59)
        at org.eclipse.aether.connector.basic.BasicRepositoryConnector$GetTaskRunner.runTask(BasicRepositoryConnector.java:447)
        at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:350)
        at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:67)
        at org.eclipse.aether.connector.basic.BasicRepositoryConnector$DirectExecutor.execute(BasicRepositoryConnector.java:581)
        at org.eclipse.aether.connector.basic.BasicRepositoryConnector.get(BasicRepositoryConnector.java:249)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.performDownloads(DefaultArtifactResolver.java:520)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:421)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:246)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:223)
        at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom(DefaultArtifactDescriptorReader.java:320)
        at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor(DefaultArtifactDescriptorReader.java:217)
        at org.eclipse.aether.internal.impl.DefaultDependencyCollector.resolveCachedArtifactDescriptor(DefaultDependencyCollector.java:535)
        at org.eclipse.aether.internal.impl.DefaultDependencyCollector.getArtifactDescriptorResult(DefaultDependencyCollector.java:519)
        at org.eclipse.aether.internal.impl.DefaultDependencyCollector.processDependency(DefaultDependencyCollector.java:409)
        at org.eclipse.aether.internal.impl.DefaultDependencyCollector.processDependency(DefaultDependencyCollector.java:363)
        at org.eclipse.aether.internal.impl.DefaultDependencyCollector.process(DefaultDependencyCollector.java:351)
        at org.eclipse.aether.internal.impl.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:254)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:341)
        at org.springframework.boot.cli.compiler.grape.AetherGrapeEngine.resolve(AetherGrapeEngine.java:302)
        at org.springframework.boot.cli.compiler.grape.AetherGrapeEngine.grab(AetherGrapeEngine.java:112)
        at org.springframework.boot.cli.compiler.grape.AetherGrapeEngine.grab(AetherGrapeEngine.java:104)
        at org.grails.cli.profile.repository.MavenProfileRepository.getAllProfiles(MavenProfileRepository.groovy:133)
        at org.grails.cli.GrailsCli.execute(GrailsCli.groovy:241)
        at org.grails.cli.GrailsCli.main(GrailsCli.groovy:162)

```

**Comment by graemerocher** [comment]
@Luxor please add the following configuration in `USER_HOME/.grails/settings.groovy`:

```
grails {
    profiles {
        repositories {
            myRepo {
               url = "http://mycompanyrepo.com"
            }
        }
    }
}

```

Where the `url` points to your internal repository, which should be configured to mirror http://repo.grails.org/grails/core/

**Comment by Luxor** [comment]
Works like a charm. Thanks you!

**Comment by israelariel** [comment]
Hello.
how can i  create my internal repository  mirroring http://repo.grails.org/grails/core/?

**Comment by graemerocher** [comment]
@israelariel You need to setup an Artifactory or Nexus mirror https://www.jfrog.com/confluence/display/RTF/Configuring+Repositories#ConfiguringRepositories-RemoteRepositories

**Comment by Schlogen** [comment]
I'm unable to reproduce this on Grails 3.1.8 as long as the profile exists in my local maven cache when the app is created. 

@weberdc Can you verify that 3.1.8 works for you as well?

**Comment by Schlogen** [comment]
Closing this. Reopen if it is happening for you on 3.1.8.

**Comment by akhalik** [comment]
I have setup grails 3.2.1. on windows JAVA_HOME and GRAILS_HOME is also set. verified by running grails -version and worked as expected.

But when issue command to create app, it always throw following error

"Error occurred running Grails CLI: unable to find valid certification path to requested target (Use --stacktrace to see the full trace)".

I tried it by disconnecting the internet and come to know its trying to connect to https://repo.grails.org

It seems that some certificates must be in JAVA_HOME/jre/lib/security/cacerts. with this thought I downloaded the SSL certificates from https://repo.grails.org by accessing the URL in browser.

same certificates imported in the JAVA_HOME/jre/lib/security/cacerts but still no change.

I could not find even a single related post over the internet.

when creating Grails app in Netbeans 8, I get the same error.

I don't know if I need any other valid certificate or there is any other problem.

Many thanks in advance for any help on this.
